modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C

```diff
@@ -128,41 +173,95 @@ ADJunctionOneToOne1PhaseUserObject::threadJoin(const UserObject & uo)
     const unsigned int c = junction_uo._connection_indices[i];
 
     _dof_indices[c] = junction_uo._dof_indices[c];
-    _solutions[c] = junction_uo._solutions[c];
+    _primitive_solutions[c] = junction_uo._primitive_solutions[c];
+    _neighbor_primitive_solutions[c] = junction_uo._neighbor_primitive_solutions[c];
     _elem_ids[c] = junction_uo._elem_ids[c];
     _local_side_ids[c] = junction_uo._local_side_ids[c];
-    _areas[c] = junction_uo._areas[c];
+    _areas_linear[c] = junction_uo._areas_linear[c];
     _directions[c] = junction_uo._directions[c];
+    _positions[c] = junction_uo._positions[c];
+    _neighbor_positions[c] = junction_uo._neighbor_positions[c];
+    _delta_x[c] = junction_uo._delta_x[c];
+    _has_neighbor[c] = junction_uo._has_neighbor[c];
   }
 }
 
 void
 ADJunctionOneToOne1PhaseUserObject::finalize()
 {
-  // Check direction compatibility
-  if (!THM::areParallelVectors(_directions[0], _directions[1]))
-    mooseError(_junction_name, ": The connected channels must be parallel at the junction.");
-
   for (unsigned int i = 0; i < _n_connections; i++)
   {
     processor_id_type owner_proc = _processor_ids[i];
+    comm().broadcast(_primitive_solutions[i], owner_proc, true);
+    comm().broadcast(_neighbor_primitive_solutions[i], owner_proc, true);
     comm().broadcast(_elem_ids[i], owner_proc, true);
     comm().broadcast(_local_side_ids[i], owner_proc, true);
-    comm().broadcast(_solutions[i], owner_proc, true);
+    comm().broadcast(_areas_linear[i], owner_proc, true);
     comm().broadcast(_directions[i], owner_proc, true);
+    comm().broadcast(_positions[i], owner_proc, true);
+    comm().broadcast(_neighbor_positions[i], owner_proc, true);
+    comm().broadcast(_delta_x[i], owner_proc, true);
+
+    // Vectors of bools are special, so we must broadcast a single bool instead
+    bool has_neighbor = _has_neighbor[i];
+    comm().broadcast(has_neighbor, owner_proc, true);
+    _has_neighbor[i] = has_neighbor;
   }
 
-  // Apply transformation to first connection's reference frame
-  const Real n_direction1 = _directions[0] * _directions[1];
-  _solutions[1][THM3Eqn::CONS_VAR_RHOUA] *= n_direction1;
+  // Check direction compatibility
+  if (!THM::areParallelVectors(_directions[0], _directions[1]))
+    mooseError(_junction_name, ": The connected channels must be parallel at the junction.");
+
+  const auto & W0_avg = _primitive_solutions[0];
+  const auto & W1_avg = _primitive_solutions[1];
 
-  _fluxes[0] = _numerical_flux.getFlux(
-      _local_side_ids[0], _elem_ids[0], true, _solutions[0], _solutions[1], _normal[0]);
+  // Multiplier for possibly different coordinate systems
+  const Real dir_mult = _directions[0] * _directions[1];
```

Comment:
Is it necessary with the assumption that the channels are parallel?

Reply:
The channels are parallel but may be in opposite directions, so `dir_mult` is equal to +/-1.

Reply:
Ok, I got confused, this cannot do 180 degree flow turn, but the direction of the flow channels could be in opposite directions.